drivers/flash/flash_esp32.c

```diff
@@ -156,6 +205,205 @@ static int flash_esp32_read(const struct device *dev, off_t address, void *buffe
 	return 0;
 }
 
+static int flash_esp32_write_check_enc(off_t address, const void *buffer, size_t length)
+{
+	int ret = 0;
+
+	if (esp_flash_encryption_enabled() && !ENCRYPTION_IS_VIRTUAL) {
+		LOG_DBG("Flash write ENCRYPTED - address 0x%lx size 0x%x", address, length);
+		ret = esp_flash_write_encrypted(NULL, address, buffer, length);
+	} else {
+		LOG_DBG("Flash write RAW - address 0x%lx size 0x%x", address, length);
+		ret = esp_flash_write(NULL, buffer, address, length);
+	}
+
+	if (ret != 0) {
+		LOG_ERR("Flash write error: %d", ret);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+#ifndef CONFIG_MCUBOOT
+
+/* Auxiliar buffer to store the sector that will be partially written */
+static uint8_t __aligned(32) write_aux_buf[FLASH_SECTOR_SIZE] = {0};
```

Comment:
Is the alignment value common across all esp chips? Any chance this magic value to receive a define to improve the readabillity?